xds/src/main/java/io/grpc/xds/internal/Matchers.java

```diff
@@ -0,0 +1,307 @@
+    @Nullable
+    public abstract Range range();
+
+    // Matches header presence.
+    @Nullable
+    public abstract Boolean present();
+
+    // Matches header value with the prefix.
+    @Nullable
+    public abstract String prefix();
+
+    // Matches header value with the suffix.
+    @Nullable
+    public abstract String suffix();
+
+    // Whether the matching semantics is inverted. E.g., present && !inverted -> !present
+    public abstract boolean inverted();
+
+    /** The request header value should exactly match the specified value. */
+    public static HeaderMatcher forExactValue(String name, String exactValue, boolean inverted) {
+      checkNotNull(name, "name");
+      checkNotNull(exactValue, "exactValue");
+      return HeaderMatcher.create(name, exactValue, null, null, null, null, null, inverted);
+    }
+
+    /** The request header value should match the regular expression pattern. */
+    public static HeaderMatcher forSafeRegEx(String name, Pattern safeRegEx, boolean inverted) {
+      checkNotNull(name, "name");
+      checkNotNull(safeRegEx, "safeRegEx");
+      return HeaderMatcher.create(name, null, safeRegEx, null, null, null, null, inverted);
+    }
+
+    /** The request header value should be within the range. */
+    public static HeaderMatcher forRange(String name, Range range, boolean inverted) {
+      checkNotNull(name, "name");
+      checkNotNull(range, "range");
+      return HeaderMatcher.create(name, null, null, range, null, null, null, inverted);
+    }
+
+    /** The request header value should exist. */
+    public static HeaderMatcher forPresent(String name, boolean present, boolean inverted) {
+      checkNotNull(name, "name");
+      return HeaderMatcher.create(name, null, null, null, present, null, null, inverted);
+    }
+
+    /** The request header value should have this prefix. */
+    public static HeaderMatcher forPrefix(String name, String prefix, boolean inverted) {
+      checkNotNull(name, "name");
+      checkNotNull(prefix, "prefix");
+      return HeaderMatcher.create(name, null, null, null, null, prefix, null, inverted);
+    }
+
+    /** The request header value should have this suffix. */
+    public static HeaderMatcher forSuffix(String name, String suffix, boolean inverted) {
+      checkNotNull(name, "name");
+      checkNotNull(suffix, "suffix");
+      return HeaderMatcher.create(name, null, null, null, null, null, suffix, inverted);
+    }
+
+    private static HeaderMatcher create(String name, @Nullable String exactValue,
+        @Nullable Pattern safeRegEx, @Nullable Range range,
+        @Nullable Boolean present, @Nullable String prefix,
+        @Nullable String suffix, boolean inverted) {
+      checkNotNull(name, "name");
+      return new AutoValue_Matchers_HeaderMatcher(name, exactValue, safeRegEx, range, present,
+          prefix, suffix, inverted);
+    }
+
+    /** Represents an integer range. */
+    @AutoValue
+    public abstract static class Range {
```

Comment:
Please follow the style guide to organize methods and nested classes properly. Definition of nested class should be put after all the other methods of the outer class.